pynicotine/gtkgui/userbrowse.py

```diff
@@ -688,34 +712,277 @@ def on_select_dir(self, selection):
 
         self.set_directory(iterator.user_data)
 
+    def on_select_file(self, selection):
+
+        if selection is None:
+            return
+
+        self.num_selected_files = selection.count_selected_rows()
+
     def on_folder_key_press_event(self, *args):
 
         keyval, keycode, state, widget = get_key_press_event_args(*args)
+        path, _focus_column = self.FolderTreeView.get_cursor()
+
+        # Note: The "Return" key is handled by on_folder_row_activated
+
+        # Shift bindings need to be above the non-shifted one
+        keycodes, mods = parse_accelerator("<Shift>Right")
+
+        if state & mods and keycode in keycodes:
+            if int(len(self.file_store)) > 0:
+                # Shift focus across to files
+                self.FileTreeView.grab_focus()
+                return True
+
+        # Expand folder node with arrow cursor
         keycodes, mods = parse_accelerator("Right")
 
-        if keycode in keycodes:
-            path, _focus_column = self.FolderTreeView.get_cursor()
+        if keycode in keycodes or (int(keyval) == 61):  # Laptop unshifted + "=" key
             self.FolderTreeView.expand_row(path, False)
             return True
 
+        # Collapse folder node with arrow cursor
         keycodes, mods = parse_accelerator("Left")
 
         if keycode in keycodes:
-            path, _focus_column = self.FolderTreeView.get_cursor()
-            self.FolderTreeView.collapse_row(path)
+            if path is not None:  # Cannot collapse a root that is already collapsed
+                self.FolderTreeView.collapse_row(path)
```

Comment:
The variable assignment is moved to the top of the definition, because 'path' is needed for many of the other functions.

The code that follows is long-winded and scruffy, I intend to find more elegant solutions for testing the conditions, ideally a way of parsing the accelerator on one line, and testing for overlapping scenarios more dynamically.

Reply:
> I intend to find more elegant solutions for testing the conditions

This is fine, but please don't include any additional large changes in this PR.

Reply:
Sure, this is already a larger fix than I had intended. Please feel free to make or suggest changes at any point in the future as you see fit.